libpod/runtime.go

```diff
@@ -562,18 +562,31 @@ func (r *Runtime) Shutdown(force bool) error {
 
 	r.valid = false
 
-	// Shutdown all containers if --force is given
-	if force {
-		ctrs, err := r.state.AllContainers()
-		if err != nil {
-			logrus.Errorf("Error retrieving containers from database: %v", err)
-		} else {
-			for _, ctr := range ctrs {
-				if err := ctr.StopWithTimeout(r.config.Engine.StopTimeout); err != nil {
-					logrus.Errorf("Error stopping container %s: %v", ctr.ID(), err)
-				}
+	ctrs, err := r.state.AllContainers()
+	if err != nil {
+		return errors.Wrap(err, "retrieve containers from database")
+	}
+	hasRunningContainers := false
+	for _, ctr := range ctrs {
+		// Shutdown all containers if --force is given
+		if force {
+			if err := ctr.StopWithTimeout(r.config.Engine.StopTimeout); err != nil {
+				logrus.Errorf("Error stopping container %s: %v", ctr.ID(), err)
 			}
 		}
+
+		state, err := ctr.State()
+		if err != nil {
+			logrus.Errorf("Unable to retrieve container state: %v", err)
+		}
+		if state == define.ContainerStateRunning {
+			hasRunningContainers = true
+		}
+	}
+	if !hasRunningContainers {
+		if err := stopPauseProcess(); err != nil {
```

Comment:
Isn't this opening us up for a new race condition?  What happens if I kill the pause process while another container is starting up and joining the pause process?

Reply:
Isn't this covered by the runtime lock in Line 556?

Reply:
That lock is purely single-process - it's basically just for mutable state within a single instance of Libpod, it does not give us any guarantee there isn't another Podman process in the process of starting a container.

Reply:
yeah that is the reason why it is kept running :(  it can potentially open a race condition and two containers end up in two differerent user namespaces.

It would be nice if systemd could keep the fd available for us, without needing any pause process.

The idea would be to pass the namespace file descriptor to the systemd user instance and not require a pause process (at least when systemd is available) at all.

I've opened an issue here: https://github.com/systemd/systemd/issues/16318

Reply:
Okay sounds good thank you for the clarification.